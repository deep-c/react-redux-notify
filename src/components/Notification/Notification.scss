$text-color-info: #5293d8;
$text-color-success: #1fe977;
$text-color-warning: #f78a46;
$text-color-error: #c00;
$background-color-info: #0098ff;
$background-color-success: #17ca65;
$background-color-warning: #ff4800;
$background-color-error: #c00;
$width: 300px;
$max-height: 150px;

@mixin notification($txt, $bg) {
  position: relative;
  word-wrap: break-word;
  pointer-events: auto;
  width: $width;
  border-radius: 5px;
  background: lighten($bg, 25%);
  padding-left: 30px;
  overflow: hidden;
  z-index: 9999;
  margin-bottom: 2px;
  max-height: $max-height;

  .icon {
    color: lighten($txt, 36%);
    background: $bg;  
  }

  .content {
    color: darken($txt, 40%);
    background-color: lighten($bg, 25%);
  }

  .actionBtn {
    color: fade-out(darken($txt, 40%), 0.4);
    &:hover {
      color: darken($txt, 40%);
    }    
  }

  .close-all {
    border: 1px solid fade-out(darken($txt, 40%), 0.7);
    color: fade-out(darken($txt, 40%), 0.4);
    &:hover {
      border-color: fade-out(darken($txt, 40%), 0.2);
      color: darken($txt, 40%);
    }
  }  
}

.icon {
  display: block;
  position: absolute;
  top: 0;
  left: 0;
  height: 100%;
  padding-top: 10px;
  box-sizing: border-box;
  width: 30px;
  text-align: center;  
}

.content {
  border-radius: 0 2px 2px 0;
}

.item {
  padding: 10px;
  font-weight: 900;
  font-family: 'Lato', sans-serif;
  border-top: 1px solid hsla(0,0%,0%,.1);
  -webkit-font-smoothing: antialiased;
  &--message {
    @extend .item;
    font-size: 0.9rem;
    max-height: $max-height;
    overflow-y: auto;
    border-top: none;
    p:last-child {
      margin-bottom: 0;
    }
  }
  &--btnBar {
    @extend .item;
    margin-top: 10px;
    padding: 0;
  }
}

.actionBtn {
  box-sizing: border-box;
  background: none;
  font-size: 0.9rem;
  font-family: 'Lato', sans-serif;
  border-right: 1px solid rgba(0, 0, 0, 0.1);  
  text-shadow: none;
  cursor: pointer;
  display: inline-block;
  width: 50%;
  height: 30px;
  line-height: 30px;
  text-align: center;

  &:hover {
    background: none;
  }

  &:last-child {
    border-right: none;
  }

  i {
    margin-right: 7px;
  }
}

.close-all {
  position: absolute;
  top: 7px;
  right: 38px;
  background: none;
  font-size: 0.9rem;
  font-family: 'Lato', sans-serif;
  text-shadow: none;
  padding: 1px 5px;
  border-radius: 3px;
  cursor: pointer;

  &:hover {
    background: none;
  }
}

.close {
  position: absolute;
  top: 0;
  right: 0;
  width: 38px;
  height: 38px;
  font-size: 0.7rem;
  line-height: 38px;
  text-align: center;
  text-shadow: none;
  color: black;
  opacity: .4;
  cursor: pointer;
  &:hover, &:focus {
    opacity: 1;
  }
  &:active {
    opacity: .2;
  }
  &:before {
    content: "\274C";
    margin: 0;
  }
}

.no-close .item--message {
  padding-right: 10px
}

.has-close-all--noDismiss .close-all {
  @extend .close-all;
  right: 8px;
}

.has-close-all--noDismiss .item--message {
  padding-right: 10px + 71px;
}

.has-close-all .item--message {
  padding-right: 10px * 2 + 95px; // space for icon and button
}

.has-close .item--message {
  padding-right: 10px + 24px; // space for icon
}

.notification--info {
  @include notification($text-color-info, $background-color-info);
}

.notification--success {
  @include notification($text-color-success, $background-color-success);
}

.notification--warning {
  @include notification($text-color-warning, $background-color-warning);
}

.notification--error {
  @include notification($text-color-error, $background-color-error);
}
