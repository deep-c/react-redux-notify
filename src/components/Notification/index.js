import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import styleMap from './Notification.scss';


export class Notification extends React.PureComponent {

  static propTypes = {
    id: PropTypes.number.isRequired,
    type: PropTypes.string.isRequired,
    canDismiss: PropTypes.bool.isRequired,
    duration: PropTypes.number.isRequired,
    icon: PropTypes.node,
    customStyles: PropTypes.object,
    customComponent: PropTypes.element,
    acceptBtn: PropTypes.shape({
      handler: PropTypes.func.isRequired,
      icon: PropTypes.node,
      title: PropTypes.node,
    }),
    denyBtn: PropTypes.shape({
      handler: PropTypes.func.isRequired,
      icon: PropTypes.node,
      title: PropTypes.node,
    }),
    isFirst: PropTypes.bool.isRequired,
    handleDismiss: PropTypes.func.isRequired,
    handleDismissAll: PropTypes.func.isRequired,
    styles: PropTypes.object.isRequired,
  }

  static defaultProps = {
    styles: styleMap,
    canDismiss: true,
    duration: 0,
  }

  componentDidMount() {
    const { handleDismiss, id, duration } = this.props;
    if (duration !== 0) {
      setTimeout(() => { handleDismiss(id); }, duration);
    }
  }

  render() {
    const { handleDismiss, handleDismissAll, isFirst, message, type, canDismiss,
      acceptBtn, denyBtn, icon, customStyles, id } = this.props;
    let { styles } = this.props;
    styles = Object.assign({}, styles, customStyles);
    const cx = classNames.bind(styles);
    const containerTypeClass = cx({
      'has-close': !isFirst && canDismiss,
      'no-close': !isFirst && !canDismiss,
      'has-close-all': isFirst && canDismiss,
      'has-close-all--noDismiss': isFirst && !canDismiss,
      [`notification--${type.toLowerCase()}`]: true,
    });

    return (
      <div key={id} className={containerTypeClass}>
        {(icon) ? <span className={styles.icon}>{icon}</span> : false}
        <div className={styles.content}>
          <div className={styles['item--message']}>{message}</div>
          { (!canDismiss && (acceptBtn || denyBtn)) ?
              <div className={styles['item--btnBar']}>
                {(acceptBtn) ?
                  <div className={styles.actionBtn}
                    onClick={(e) => {
                      acceptBtn.handler(e, this.props);
                    }}>
                    {(acceptBtn.icon && typeof acceptBtn.icon === 'string') ?
                      <i className={acceptBtn.icon} />
                      :
                      acceptBtn.icon
                    }
                    {acceptBtn.title}
                  </div>
                  : false
                }
                {(denyBtn) ?
                  <div className={styles.actionBtn}
                    onClick={(e) => {
                      denyBtn.handler(e, this.props);
                    }}>
                    {(denyBtn.icon && typeof denyBtn.icon === 'string') ?
                      <i className={denyBtn.icon} />
                      :
                      denyBtn.icon
                    }
                    {denyBtn.title}
                  </div>
                  :
                  false
                }
              </div>
              :
              false
          }
        </div>
        {(canDismiss) ?
          <div className={styles.close} onClick={() => handleDismiss(id)}></div>
          :
          false
        }
        {(isFirst && canDismiss) ?
          <div className={styles['close-all']}
            onClick={() => handleDismissAll()}>Close All</div>
          : false
        }
      </div>
    );
  }
}

export default Notification;
