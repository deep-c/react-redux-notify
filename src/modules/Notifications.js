// Constants
export const NOTIFICATIONS_POS_TOP_RIGHT = 'TopRight';
export const NOTIFICATIONS_POS_BOT_RIGHT = 'BottomRight';
export const NOTIFICATIONS_POS_BOT_LEFT = 'BottomLeft';
export const NOTIFICATIONS_POS_TOP_LEFT = 'TopLeft';
export const NOTIFICATION_TYPE_SUCCESS = 'SUCCESS';
export const NOTIFICATION_TYPE_WARNING = 'WARNING';
export const NOTIFICATION_TYPE_INFO = 'INFO';
export const NOTIFICATION_TYPE_ERROR = 'ERROR';
export const ADD_NOTIFICATION = 'ADD_NOTIFICATION';
export const REMOVE_NOTIFICATION = 'REMOVE_NOTIFICATION';
export const REMOVE_ALL_NOTIFICATIONS = 'REMOVE_ALL_NOTIFICATIONS';

// Action Creators
export function createNotification(notification) {
  return {
    type: ADD_NOTIFICATION,
    notification: {
      id: parseInt(Math.random().toString().split('.')[1], 10),
      ...notification,
    },
  };
}

export function removeNotification(id) {
  return {
    type: REMOVE_NOTIFICATION,
    id,
  };
}

export function removeAllNotifications(force) {
  return {
    type: REMOVE_ALL_NOTIFICATIONS,
    force,
  };
}

// Reducer
export const initialState = [];
export default function (state = initialState, action) {
  switch (action.type) {
    case ADD_NOTIFICATION:
      return [...state, action.notification];
    case REMOVE_NOTIFICATION:
      return state.filter(item => item.id !== action.id);
    case REMOVE_ALL_NOTIFICATIONS:
      if (action.force) {
        return [];
      }
      return state.filter(item => !item.canDismiss);
    default:
      return state;
  }
}
