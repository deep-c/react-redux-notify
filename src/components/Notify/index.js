import React from 'react';
import { connect } from 'react-redux';
import ReactCSSTransitionGroup from 'react-addons-css-transition-group';
import PureRenderMixin from 'react-addons-pure-render-mixin';
import { removeNotification, removeAllNotifications, NOTIFICATIONS_POS_TOP_RIGHT } from 'modules/Notifications';
import { default as Notification } from 'components/Notification';
import styleMap from './Notify.scss';


export class Notify extends React.Component {

  static propTypes = {
    notifications: React.PropTypes.array.isRequired,
    remove: React.PropTypes.func.isRequired,
    removeAll: React.PropTypes.func.isRequired,
    styles: React.PropTypes.object.isRequired,
    customStyles: React.PropTypes.object,
    notificationComponent: React.PropTypes.func,
    transitionDurations: React.PropTypes.shape({
      enter: React.PropTypes.number,
      leave: React.PropTypes.number,
    }),
    position: React.PropTypes.string,
    forceClose: React.PropTypes.bool,
  }

  static defaultProps = {
    notificationComponent: Notification,
    transitionDurations: {
      enter: 160,
      leave: 400,
    },
    position: NOTIFICATIONS_POS_TOP_RIGHT,
    styles: styleMap,
    forceClose: false,
  }

  constructor(props) {
    super(props);
    this.shouldComponentUpdate = PureRenderMixin.shouldComponentUpdate.bind(this);
    this.handleDismiss = this._handleDismiss.bind(this);
    this.handleDismissAll = this._handleDismissAll.bind(this);
  }

  _handleDismiss(id) {
    const { remove } = this.props;
    remove(id);
  }

  _handleDismissAll(force) {
    const { removeAll, forceClose } = this.props;
    removeAll(force || forceClose);
  }

  render() {
    const {
      notifications,
      customStyles,
      notificationComponent,
      transitionDurations,
      position } = this.props;
    let { styles } = this.props;
    styles = Object.assign({}, styles, customStyles);
    const notificationsContainerClass = styles[`container${position}`];

    return (
        <div className={notificationsContainerClass}>
          <ReactCSSTransitionGroup
              component="div"
              className={styles.wrapper}
              transitionName={ { enter: styles.enter,
                                  leave: styles.leave,
                              } }
              transitionEnterTimeout={transitionDurations.enter}
              transitionLeaveTimeout={transitionDurations.leave}>
            {
              notifications.map((notification, i) => {
                const NewNotification = notification.customComponent || notificationComponent;
                return (
                  <NewNotification
                    key={notification.id}
                    {...notification}
                    isFirst={(i === 0 && notifications.length > 1)}
                    handleDismiss={this.handleDismiss}
                    handleDismissAll={this.handleDismissAll}
                  />
                );
              })
            }
          </ReactCSSTransitionGroup>
        </div>
    );
  }
}

const mapStateToProps = state => ({
  notifications: state.notifications,
});

const mapDispatchToProps = dispatch => ({
  remove: (id) => {
    dispatch(removeNotification(id));
  },
  removeAll: (force) => {
    dispatch(removeAllNotifications(force));
  },
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Notify);
