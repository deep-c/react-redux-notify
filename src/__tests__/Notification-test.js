import React from 'react';
import { shallow } from 'enzyme';
import { Notification } from 'components/Notification';
import { NOTIFICATION_TYPE_SUCCESS } from 'modules/Notifications';

describe('Notification', () => {
  const ID = Date.now();
  const handleDismissClick = jest.fn();
  const handleDismissAllClick = jest.fn();
  const props = {
    id: ID,
    message: 'Hello There!',
    type: NOTIFICATION_TYPE_SUCCESS,
    isFirst: false,
    duration: 0,
    handleDismiss: handleDismissClick,
    handleDismissAll: handleDismissAllClick,
    localization: {
      closeAllBtnText: 'Close All',
      acceptBtnText: 'Accept',
      denyBtnText: 'Deny',
    },
    customStyles: {},
  };

  it('renders with default props', () => {
    const component = shallow(<Notification {...props} />);
    expect(component).toMatchSnapshot();
  });

  it('renders with an element for the icon when an element is passed', () => {
    const tProps = { ...props, icon: <i className="fa fa-fire" /> };
    const component = shallow(<Notification {...tProps} />);
    expect(component).toMatchSnapshot();
  });

  it('renders with an string for the icon when a string is passed', () => {
    const tProps = { ...props, icon: String.fromCharCode(183) };
    const component = shallow(<Notification {...tProps} />);
    expect(component).toMatchSnapshot();
  });

  it('renders an accept button with an a title and an icon className and calls function on click', () => {
    const tProps = {
      ...props,
      canDismiss: false,
      acceptBtn: {
        handler: jest.fn(),
        icon: 'fa fa-thumbs-up',
        title: 'Accept',
      },
    };
    const component = shallow(<Notification {...tProps} />);
    expect(component).toMatchSnapshot();
  });

  it('renders an accept button without a title and an icon element ', () => {
    const tProps = {
      ...props,
      canDismiss: false,
      acceptBtn: {
        handler: jest.fn(),
        icon: <i className="fa fa-thumbs-up" />,
      },
    };
    const component = shallow(<Notification {...tProps} />);
    expect(component).toMatchSnapshot();
  });

  it('renders a deny button with an a title and an icon className', () => {
    const tProps = {
      ...props,
      canDismiss: false,
      denyBtn: {
        handler: jest.fn(),
        icon: 'fa fa-thumbs-down',
        title: 'Deny',
      },
    };
    const component = shallow(<Notification {...tProps} />);
    expect(component).toMatchSnapshot();
  });

  it('renders a deny button without a title and an icon element ', () => {
    const tProps = {
      ...props,
      canDismiss: false,
      denyBtn: {
        handler: jest.fn(),
        icon: <i className="fa fa-thumbs-down" />,
      },
    };
    const component = shallow(<Notification {...tProps} />);
    expect(component).toMatchSnapshot();
  });

  it('renders with a close all button', () => {
    const tProps = { ...props, isFirst: true };
    const component = shallow(<Notification {...tProps} />);
    expect(component.find('.close-all').text()).toBe('Close All');
    expect(component).toMatchSnapshot();
  });

  it('renders with a duration greater than 0', () => {
    const tProps = { ...props, duration: 2000 };
    const component = shallow(<Notification {...tProps} />);
    expect(component).toMatchSnapshot();
  });

  it('calls handleDismiss after duration expires', () => {
    jest.useFakeTimers();
    const tProps = { ...props, duration: 2000 };
    // eslint-disable-next-line no-unused-vars
    const component = shallow(<Notification {...tProps} />);
    setTimeout(() => {
      expect(handleDismissClick).toHaveBeenCalled();
    }, 2000);
    jest.runAllTimers();
  });

  it('calls the button handler function when clicked', () => {
    const acceptBtnHandler = jest.fn();
    const denyBtnHandler = jest.fn();
    const tProps = {
      ...props,
      canDismiss: false,
      denyBtn: {
        handler: denyBtnHandler,
        icon: <i className="fa fa-thumbs-down" />,
        title: 'Deny',
      },
      acceptBtn: {
        handler: acceptBtnHandler,
        icon: 'fa fa-thumbs-up',
        title: 'Accept',
      },
    };
    const component = shallow(<Notification {...tProps} />);
    const buttons = component.find('.actionBtn');
    const acceptBtn = buttons.first();
    const denyBtn = buttons.last();
    acceptBtn.simulate('click');
    expect(acceptBtnHandler).toBeCalledWith(undefined, tProps);
    denyBtn.simulate('click');
    expect(denyBtnHandler).toBeCalledWith(undefined, tProps);
  });

  it('calls handleDismiss onclick', () => {
    const component = shallow(<Notification {...props} />);
    component.find('.close').simulate('click');
    expect(handleDismissClick).toHaveBeenCalled();
  });

  it('calls handleDismissAll onclick', () => {
    const tProps = { ...props, isFirst: true };
    const component = shallow(<Notification {...tProps} />);
    component.find('.close-all').simulate('click');
    expect(handleDismissAllClick).toHaveBeenCalled();
  });

  it('renders custom localization text for accept and deny buttons', () => {
    const acceptBtnHandler = jest.fn();
    const denyBtnHandler = jest.fn();
    const tProps = {
      ...props,
      localization: {
        closeAllBtnText: 'Close All',
        acceptBtnText: 'Yes',
        denyBtnText: 'No',
      },
      canDismiss: false,
      denyBtn: {
        handler: denyBtnHandler,
        icon: false,
      },
      acceptBtn: {
        handler: acceptBtnHandler,
        icon: false,
      },
      isFirst: true,
    };
    const component = shallow(<Notification {...tProps} />);
    const buttons = component.find('.actionBtn');
    expect(buttons.first().text()).toBe('Yes');
    expect(buttons.last().text()).toBe('No');
    expect(component).toMatchSnapshot();
  });

  it('renders custom localization text for close all', () => {
    const tProps = {
      ...props,
      localization: {
        closeAllBtnText: 'Close dem all',
        acceptBtnText: 'Yes',
        denyBtnText: 'No',
      },
      isFirst: true,
    };
    const component = shallow(<Notification {...tProps} />);
    expect(component.find('.close-all').text()).toBe('Close dem all');
    expect(component).toMatchSnapshot();
  });
});
