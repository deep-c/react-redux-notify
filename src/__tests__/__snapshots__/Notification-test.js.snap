// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notification renders a deny button with an a title and an icon className 1`] = `
<div
  className="no-close notification--success"
>
  <div
    className="content"
  >
    <div
      className="item__message"
    >
      Hello There!
    </div>
    <div
      className="item__btnBar"
    >
      <div
        className="actionBtn"
        onClick={[Function]}
      >
        <i
          className="fa fa-thumbs-down"
        />
        Deny
      </div>
    </div>
  </div>
</div>
`;

exports[`Notification renders a deny button without a title and an icon element  1`] = `
<div
  className="no-close notification--success"
>
  <div
    className="content"
  >
    <div
      className="item__message"
    >
      Hello There!
    </div>
    <div
      className="item__btnBar"
    >
      <div
        className="actionBtn"
        onClick={[Function]}
      >
        <i
          className="fa fa-thumbs-down"
        />
        Deny
      </div>
    </div>
  </div>
</div>
`;

exports[`Notification renders an accept button with an a title and an icon className and calls function on click 1`] = `
<div
  className="no-close notification--success"
>
  <div
    className="content"
  >
    <div
      className="item__message"
    >
      Hello There!
    </div>
    <div
      className="item__btnBar"
    >
      <div
        className="actionBtn"
        onClick={[Function]}
      >
        <i
          className="fa fa-thumbs-up"
        />
        Accept
      </div>
    </div>
  </div>
</div>
`;

exports[`Notification renders an accept button without a title and an icon element  1`] = `
<div
  className="no-close notification--success"
>
  <div
    className="content"
  >
    <div
      className="item__message"
    >
      Hello There!
    </div>
    <div
      className="item__btnBar"
    >
      <div
        className="actionBtn"
        onClick={[Function]}
      >
        <i
          className="fa fa-thumbs-up"
        />
        Accept
      </div>
    </div>
  </div>
</div>
`;

exports[`Notification renders custom localization text for accept and deny buttons 1`] = `
<div
  className="has-close-all--noDismiss notification--success"
>
  <div
    className="content"
  >
    <div
      className="item__message"
    >
      Hello There!
    </div>
    <div
      className="item__btnBar"
    >
      <div
        className="actionBtn"
        onClick={[Function]}
      >
        Yes
      </div>
      <div
        className="actionBtn"
        onClick={[Function]}
      >
        No
      </div>
    </div>
  </div>
</div>
`;

exports[`Notification renders custom localization text for close all 1`] = `
<div
  className="has-close-all notification--success"
>
  <div
    className="content"
  >
    <div
      className="item__message"
    >
      Hello There!
    </div>
  </div>
  <div
    className="close"
    onClick={[Function]}
  />
  <div
    className="close-all"
    onClick={[Function]}
  >
    Close dem all
  </div>
</div>
`;

exports[`Notification renders with a close all button 1`] = `
<div
  className="has-close-all notification--success"
>
  <div
    className="content"
  >
    <div
      className="item__message"
    >
      Hello There!
    </div>
  </div>
  <div
    className="close"
    onClick={[Function]}
  />
  <div
    className="close-all"
    onClick={[Function]}
  >
    Close All
  </div>
</div>
`;

exports[`Notification renders with a duration greater than 0 1`] = `
<div
  className="has-close notification--success"
>
  <div
    className="content"
  >
    <div
      className="item__message"
    >
      Hello There!
    </div>
  </div>
  <div
    className="close"
    onClick={[Function]}
  />
</div>
`;

exports[`Notification renders with an element for the icon when an element is passed 1`] = `
<div
  className="has-close notification--success"
>
  <span
    className="icon"
  >
    <i
      className="fa fa-fire"
    />
  </span>
  <div
    className="content"
  >
    <div
      className="item__message"
    >
      Hello There!
    </div>
  </div>
  <div
    className="close"
    onClick={[Function]}
  />
</div>
`;

exports[`Notification renders with an string for the icon when a string is passed 1`] = `
<div
  className="has-close notification--success"
>
  <span
    className="icon"
  >
    Â·
  </span>
  <div
    className="content"
  >
    <div
      className="item__message"
    >
      Hello There!
    </div>
  </div>
  <div
    className="close"
    onClick={[Function]}
  />
</div>
`;

exports[`Notification renders with default props 1`] = `
<div
  className="has-close notification--success"
>
  <div
    className="content"
  >
    <div
      className="item__message"
    >
      Hello There!
    </div>
  </div>
  <div
    className="close"
    onClick={[Function]}
  />
</div>
`;
